corehq/apps/app_manager/static/app_manager/js/hq.app_data_drilldown.ng.js

```diff
@@ -0,0 +1,248 @@
+            util.setModules(module_choices);
+            $scope.selectedFormData = {};
+            util.setForms();
+        };
+
+        $scope.updateFormChoices = function () {
+            if ($scope.formData.application && $scope.formData.module) {
+                var form_choices = self._forms_by_app_by_module[$scope.formData.application][$scope.formData.module];
+                util.setForms(form_choices);
+            }
+        };
+
+        $scope.updateCaseTypeChoices = function () {
+            var case_type_choices = self._case_types_by_app[$scope.formData.application];
+            util.setCaseTypes(case_type_choices);
+            $scope.hasNoCaseTypes = _.isEmpty(case_type_choices);
+        };
+
+        $scope.handleSubmitForm = function () {
+            $scope.isSubmittingForm = true;
+
+            // Immediately copy form data object so that if modal is
+            // accidentally dismissed, data is still properly sent
+            var formData = _.clone($scope.formData);
+
+            djangoRMI[djangoRMICallbackName]({
+                formData: formData
+            })
+            .success(function (data) {
+                if (data.success) {
+                    processApplicationDataFormSuccessCallback(data);
+                } else {
+                    $scope.isSubmittingForm = false;
+                    $scope.formSubmissionError = data.error;
+                }
+            })
+            .error(function () {
+                $scope.isSubmittingForm = false;
+                $scope.formSubmissionError = 'default';
+            });
+        };
+
+        $scope.selectedAppData = {};
+        $scope.$watch('formData.application', function (curApp) {
+            var currentApps = _.filter(self._apps_by_type[$scope.formData.app_type], function (app) {
+                return app.id === curApp;
+            });
+            if (currentApps.length === 0) {
+                $scope.selectedAppData = {};
+            } else {
+                var app_data = currentApps[0].data;
+                app_data.name = currentApps[0].text;
+                $scope.selectedAppData = app_data;
+            }
+        });
+
+        $scope.selectedFormData = {};
+        $scope.$watch('formData.form', function (curForm) {
+            if ($scope.formData.application && $scope.formData.module) {
+                var curForms = _.filter(self._forms_by_app_by_module[$scope.formData.application][$scope.formData.module], function (form) {
+                    return form.id === curForm;
+                });
+                if (curForms.length === 0) {
+                    $scope.selectedFormData = {};
+                } else {
+                    var form_data = curForms[0].data;
+                    form_data.name = curForms[0].text;
+                    $scope.selectedFormData = form_data;
+                }
+            }
+        });
+    };
+    app_drilldown.controller(selectControllers);
+
+}(window.angular));
```

Comment:
in general would be nice to see tests for these. this logic is complex and might break in the future